include/boost/geometry/util/promote_integral.hpp

```diff
@@ -0,0 +1,156 @@
+// Boost.Geometry (aka GGL, Generic Geometry Library)
+
+// Copyright (c) 2015, Oracle and/or its affiliates.
+
+// Contributed and/or modified by Menelaos Karavelas, on behalf of Oracle
+
+// Licensed under the Boost Software License version 1.0.
+// http://www.boost.org/users/license.html
+
+#ifndef BOOST_GEOMETRY_UTIL_PROMOTE_INTEGRAL_HPP
+#define BOOST_GEOMETRY_UTIL_PROMOTE_INTEGRAL_HPP
+
+// For now deactivate the use of multiprecision integers
+// TODO: activate it later
+#define BOOST_GEOMETRY_NO_MULTIPRECISION_INTEGER
+
+#include <climits>
+
+#include <boost/mpl/empty.hpp>
+#include <boost/mpl/front.hpp>
+#include <boost/mpl/if.hpp>
+#include <boost/mpl/list.hpp>
+#include <boost/mpl/pop_front.hpp>
+
+#if !defined(BOOST_GEOMETRY_NO_MULTIPRECISION_INTEGER)
+#include <boost/multiprecision/cpp_int.hpp>
+#endif
+
+#include <boost/type_traits/is_integral.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace promote_integral
+{
+
+template
+<
+    typename T,
+    typename List,
+    bool IsEmpty = boost::mpl::empty<List>::type::value
+>
+struct promote_to_larger
+{
+    typedef typename boost::mpl::if_c
+        <
+            (sizeof(typename boost::mpl::front<List>::type) >= 2 * sizeof(T)),
+            typename boost::mpl::front<List>::type,
+            typename promote_to_larger
+                <
+                    T, typename boost::mpl::pop_front<List>::type
+                >::type
+        >::type type;
```

Comment:
Wow, I haven't expected using MPL for this, but ok it's flexible.

Why not "iterate" over the MPL sequence instead of "popping"? Wouldn't it take less time to compile?

Is the `sizeof(boost::multiprecision::number<..., 2 * CHAR_BIT * sizeof(T), ...>)` always `>=` than `2*sizeof(T)`? Because if not, `bm::number<>` won't be picked.